## Universidad de los Andes

## Facultad de Ingeniería

Departamento de Ingeniería Industrial

Maestría en Inteligencia Analítica de Datos

Materia: Despliegue de Soluciones Analíticas

Taller Grupal 3 Despliegue de Soluciones Analíticas

Bogotá, noviembre de 2025

Estudiantes:
*   Diego Alejandro Lemus Guzmán
*   Valeria Iglesias Miranda
*   Sergio Andrés Perdomo Murcia
*   Danilo Suárez Vargas

## Taller Grupal 3 Despliegue de Soluciones Analíticas - Exploracion de la base de datos del proyecto

# Objetivo
Realizar la limpieza de los datos, realizando transformaciones necesarias antes de entrenar el modelo.


In [50]:
# Librerias a emplear
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [51]:
# Primeras filas del DataFrame, información general (tipos de datos, número de filas y columnas)
# estadísticas descriptivas de las columnas numéricas.

df_anuncios = pd.read_csv('data_clean_anuncios.csv')

display(df_anuncios.head())
display(df_anuncios.info())
display(df_anuncios.describe())

anuncio  \
0                                                                                                                                                     Desde $ 223.000.000\n\nLocal Comercial en Bello, Antioquia\n\nUnidades desde:\n\n1 Ambiente\n10.88 m²\nPlaza Beillin, Local En Venta En Bello\n\nSomos el lugar donde el comercio, el bienestar y los servicios esenciales convergen para ofrecerte una experiencia que hace parte de tu día a día. Diseñamos un espacio pensado para facilitar tu vida, donde cada rincón refleja la calidez y el dinamismo de nuestra ciudad.\n\nContactar\nLlamar\nWhatsapp   
1                                                                                                                                                                                              Desde $ 683.600.000\n\nApartamento en El Retiro, Antioquia\n\nUnidades desde:\n\n1 Hab\n2 Baños\n55.02 m²\nEntrelagos Apartamentos El Retiro, Apartamento En Venta En El Retiro\n\n¡Nuevo proyecto! Apartamentos de 1, 2 y 3 alcobas ubicado en el Retiro - Sector La Argentina. Un proyecto consciente rodeado de encanto natural, donde el lago será el centro de las zonas comunes.\n\nContactar\nLlamar\nWhatsapp   
2                                                                                                                                                                                                       Desde $ 540.000.000\n\nLote en La ceja, Antioquia\n\nUnidades desde:\n\n1.246 m²\nCataluña Condomino, Lote En Venta En La Ceja\n\nLotes urbanizados desde 1.000m2 hasta 1969m2 desde 480.000.000 millones en adelante para entrega en mayo 2025, un proyecto inspirado en el deseo de llevar una vida más natural y orgánica, que desde su concepción respete la naturalezas.\n\nContactar\nLlamar\nWhatsapp   
3  Desde $ 243.000.000\n\nApartamento en Caldas, Antioquia\n\nUnidades desde:\n\n2 Habs.\n2 Baños\n50.9 m²\nSonora Montana, Apartamento En Venta En Caldas\n\nDescubre SONORA MONTANA, el primer paso hacia un estilo de vida conectado con la naturaleza en SONORA VALLE RESIDENCIAL, proyecto VIS con dos habitaciones o posibilidad de 3, áreas . Ubicado en Caldas, Antioquia, este proyecto redefine la armonía entre la vida en ciudad y la tranquilidad campestre. Se encuentra estratégicamente cerca de la Universidad La Sallista y ofrece un fácil acceso a través de la Variante de Caldas.\n\nContactar   
4                                                                      Desde $ 245.000.000\n\nApartaestudio en Caldas, Antioquia\n\nUnidades desde:\n\n2 Habs.\n2 Baños\n52.41 m²\nTierra Dulce, Apartaestudio En Venta En Caldas\n\nTierra Dulce, Proyecto VIS en lanzamiento Ubicado en Caldas Antioquia sobre toda la variante, 52 metros diseñados para tu comodidad, paga hasta el 2027 tu cuota inicial y haz realidad tu sueño de tener casa propia !! Aplica Subsidios. Cómodos espacios 2 habitaciones más estudio o 3 habitaciones, escribenos para darte más información !\n\nContactar\nLlamar\nWhatsapp   

   valor_cop           tipo  \
0  223000000          Local   
1  683600000    Apartamento   
2  540000000           Lote   
3  243000000    Apartamento   
4  245000000  Apartaestudio   

                                                                                           city  \
0                           Antioquia\n\nUnidades desde:\n\n1 Ambiente\n10.88 m²\nPlaza Beillin   
1   Antioquia\n\nUnidades desde:\n\n1 Hab\n2 Baños\n55.02 m²\nEntrelagos Apartamentos El Retiro   
2                                  Antioquia\n\nUnidades desde:\n\n1.246 m²\nCataluña Condomino   
3                     Antioquia\n\nUnidades desde:\n\n2 Habs.\n2 Baños\n50.9 m²\nSonora Montana   
4                      Antioquia\n\nUnidades desde:\n\n2 Habs.\n2 Baños\n52.41 m²\nTierra Dulce   

      Ciudad                             Estado    area  Cuartos  bano  \
0      Bello            Local En Venta En Bello  10.880        8   NaN   
1  El Retiro  Apartamento En Venta En El Retiro  55.020   

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 657 entries, 0 to 656
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   anuncio       657 non-null    object 
 1   valor_cop     657 non-null    int64  
 2   tipo          657 non-null    object 
 3   city          653 non-null    object 
 4   Ciudad        657 non-null    object 
 5   Estado        647 non-null    object 
 6   area          657 non-null    float64
 7   Cuartos       657 non-null    int64  
 8   bano          0 non-null      float64
 9   habitaciones  657 non-null    object 
dtypes: float64(2), int64(2), object(6)
memory usage: 51.5+ KB


None

valor_cop         area     Cuartos  bano
count  6.570000e+02   657.000000  657.000000   0.0
mean   4.343351e+08    94.665595    1.958904   NaN
std    3.660674e+08   219.353644    1.086145   NaN
min    5.500000e+07     1.000000    0.000000   NaN
25%    2.171360e+08    47.140000    1.000000   NaN
50%    3.391204e+08    58.930000    2.000000   NaN
75%    5.300000e+08    76.500000    2.000000   NaN
max    3.753300e+09  2484.800000    9.000000   NaN

In [52]:
## Identificacion y gestion de datos faltantes
missing_values = df_anuncios.isnull().sum()
print("Número de valores faltantes por columna:")
print(missing_values)

missing_percentage = (missing_values / len(df_anuncios)) * 100
print("\nPorcentaje de valores faltantes por columna:")
print(missing_percentage)

Número de valores faltantes por columna:
anuncio           0
valor_cop         0
tipo              0
city              4
Ciudad            0
Estado           10
area              0
Cuartos           0
bano            657
habitaciones      0
dtype: int64

Porcentaje de valores faltantes por columna:
anuncio           0.000000
valor_cop         0.000000
tipo              0.000000
city              0.608828
Ciudad            0.000000
Estado            1.522070
area              0.000000
Cuartos           0.000000
bano            100.000000
habitaciones      0.000000
dtype: float64


## Propuestas para el manejo de valores faltantes por columna:
- 'bano': Todos los valores faltantes. Se propone eliminar esta columna.
- 'city': 0.61% valores faltantes. Se propone eliminar esta columna, dado que la columna 'Ciudad' proporciona información ya normalizada sin faltantes.
- 'Estado': 1.52% valores faltantes. Se propone eliminar esta columna, ya que la normalizacion de estos datos no ofrecerán info relevante para su uso.

In [53]:
#cuando "city" es nulo
df_anuncios[df_anuncios['city'].isna()]

#eliminar filas cuando city es nulo
df_anuncios = df_anuncios.dropna(subset=['city'])

In [54]:
#escogiendo columnas utiles para el modelo
columnas_utiles = ['valor_cop', 'area', 'Cuartos', 'tipo', 'city', 'Ciudad']
df_anuncios = df_anuncios[columnas_utiles]

### Limpieza de columna 'tipo'

In [55]:
# Pasar todo a minúsculas y eliminar espacios
df_anuncios['tipo'] = df_anuncios['tipo'].str.strip().str.lower()

df_anuncios['tipo'] = df_anuncios['tipo'].replace({
    'apartamento ': 'apartamento',
    'apto': 'apartamento',
    'apartamentoo': 'apartamento',
    'csa': 'casa'
})

#filtrar solo por apartamentos y casas, ya que para el resto de inmuebles no se tiene suficiente información
df_anuncios = df_anuncios[df_anuncios['tipo'].isin(['apartamento', 'casa'])]

print(df_anuncios['tipo'].value_counts())


tipo
apartamento    548
casa            68
Name: count, dtype: int64


### Limpieza columna 'city'

In [56]:
#Convertir a string por si hay valores numéricos o nulos
df_anuncios['city'] = df_anuncios['city'].astype(str)

#Pasar todo a minúsculas y eliminar espacios al inicio y final
df_anuncios['city'] = df_anuncios['city'].str.strip().str.lower()

#Eliminar saltos de línea, tabs y espacios extra
df_anuncios['city'] = df_anuncios['city'].str.replace(r'\s+', ' ', regex=True)

#Extraer solo la primera palabra antes de "Unidades" o del salto de línea original
df_anuncios['city'] = df_anuncios['city'].str.extract(r'^([\w\sáéíóúÁÉÍÓÚñÑ]+)', expand=False)  

# Eliminar espacios al inicio y final
df_anuncios['city'] = df_anuncios['city'].str.strip()

# Pasar a formato título (Antioquia → Antioquia)
df_anuncios['city'] = df_anuncios['city'].str.title()

# Extraer la parte inicial que parece un departamento o ciudad
df_anuncios['city'] = df_anuncios['city'].str.extract(
    r'^([A-Za-zÁÉÍÓÚáéíóúÑñ\s]+?)(?:\s+Unidades|\s+\d|\n|$)',
    expand=False
)

#Contar la frecuencia de cada valor
frecuencias = df_anuncios['city'].value_counts()

# Filtrar solo los valores que aparecen al menos 5 veces
valores_comunes = frecuencias[frecuencias >= 5].index

df_anuncios = df_anuncios[df_anuncios['city'].isin(valores_comunes)]

#Revisar los valores que quedaron
pd.set_option('display.max_rows', None)
print(df_anuncios['city'].value_counts())



city
Bogotá             121
Cali                65
Barranquilla        48
Cartagena           28
Santa Marta         27
Medellín            21
Cundinamarca        18
Villavicencio       15
Jamundí             15
Pereira             13
Antioquia           13
Armenia             13
Cajicá              12
Piedecuesta         12
Bello               12
Bucaramanga         11
Valle Del Cauca     11
Santander           10
Soacha               9
Zipaquirá            8
Atlantico            8
Sabaneta             7
Ibagué               6
Dosquebradas         6
La Estrella          5
Madrid               5
Floridablanca        5
Rionegro             5
Candelaria           5
Tolima               5
Soledad              5
Risaralda            5
Name: count, dtype: int64


### Limpieza de la variable ciudad

In [57]:
#Agrupando barrios poco frecuentes, como minimo tener 5 datos
umbral = 5

# Contar frecuencia de cada barrio dentro de cada ciudad
frecuencias_barrio_ciudad = df_anuncios.groupby('city')['Ciudad'].value_counts()

# Función para limpiar barrios raros
def limpiar_barrio(row):
    try:
        if frecuencias_barrio_ciudad[row['city']][row['Ciudad']] >= umbral:
            return row['Ciudad']
        else:
            return 'Otros'
    except KeyError:
        return 'Otros'  # por si hay algún valor raro
       
df_anuncios['barrio_limpio'] = df_anuncios.apply(limpiar_barrio, axis=1)

#Realizando target encoding
# Calcular promedio de precio por ciudad y barrio_limpio
target_mean = df_anuncios.groupby(['city', 'barrio_limpio'])['valor_cop'].mean()

# Mapear el encoding al DataFrame
df_anuncios['barrio_target_enc'] = df_anuncios.apply(
    lambda x: target_mean[x['city'], x['barrio_limpio']], axis=1
)

pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', None)
print(df_anuncios[['city', 'barrio_limpio']].value_counts().head(200))


city             barrio_limpio     
Bogotá           Otros                 82
Cali             Otros                 41
Cartagena        Otros                 22
Barranquilla     Otros                 21
Santa Marta      Otros                 19
Cali             Bochalema             15
Jamundí          Otros                 15
Antioquia        Otros                 13
Medellín         Otros                 13
Cajicá           Otros                 12
Bogotá           Fontibon              12
Cundinamarca     Otros                 12
Bello            Otros                 12
Barranquilla     Miramar               12
Bucaramanga      Otros                 11
Valle Del Cauca  Otros                 11
Villavicencio    Otros                 10
Bogotá           Lagos de Torca         9
Cali             Valle del Lili         9
Pereira          Otros                  8
Santa Marta      Pozos Colorados        8
Medellín         El Poblado             8
Zipaquirá        Otros                  

### Limpieza variables area y cuartos


In [58]:
# Revisión de valores nulos
print(df_anuncios[['area', 'Cuartos']].isnull().sum())

print("Valores únicos de Cuartos:", df_anuncios['Cuartos'].unique())
print("Valores únicos de area:", df_anuncios['area'].unique()[:20])  


area       0
Cuartos    0
dtype: int64
Valores únicos de Cuartos: [2 3 5 1 4 6]
Valores únicos de area: [ 55.02  50.9   62.    62.45  63.22  75.58 140.    87.35  59.8   83.5
  52.    94.    54.51  37.81  54.99  65.31 191.    42.38  59.89 600.  ]


In [62]:
print(df_anuncios.columns)


Index(['valor_cop', 'area', 'Cuartos', 'tipo', 'city', 'Ciudad',
       'barrio_limpio', 'barrio_target_enc'],
      dtype='object')


In [60]:
# Guardar en CSV
df_anuncios.to_csv('df_anuncios_procesado.csv', index=False)

# Resumen

- Para la columna **city** fue necesario eliminar los valores nulos, que correspondían al 0.6% de los datos. Se verificó que estas filas no contenían información relevante de ubicación geográfica adicional, por lo que se decidió eliminar todas las filas con valores nulos en esta columna. Posteriormente, se realizó una limpieza con expresiones regulares, eliminando espacios innecesarios y dejando solo la primera palabra o expresiones relevantes antes de un salto de línea, debido a que esta variable contenía información extra del anuncio. Se eliminaron filas donde no fue posible realizar la limpieza correctamente.

- Para la columna **tipo** se observó que los tipos "apartamento" y "casa" representaban el 95% de los registros, mientras que el resto de tipos de inmueble contaba con muy poca información. Por lo tanto, se decidió eliminar los registros correspondientes a los tipos minoritarios y trabajar únicamente con estas dos categorías principales.

- Para la columna **Ciudad**, que corresponde principalmente a la zona o barrio del inmueble, se detectó una alta variabilidad. Se procedió a agrupar los barrios por ciudad, creando una nueva columna llamada barrio_limpio, donde si un barrio tenía menos de 5 registros dentro de su ciudad se asignaba la categoría "Otros", y si tenía 5 o más registros se mantenía su valor original.

- Se creó una nueva columna **barrio_target_enc**, mediante Target Encoding, calculando el promedio de valor_cop por combinación de ciudad y barrio limpio. Esta columna permite reflejar de manera numérica la información de la ubicación del inmueble, facilitando su uso en modelos de predicción de precios.

- Para las columnas **area** y **Cuartos** no fue necesaria ninguna transformación, ya que los valores se encontraban consistentes y listos para su uso como variables numéricas en los modelos.